Add about: scheme handler.

basuke opened the pull request:
#### 63ad5df1796404e0959edb7fba6cc8344b84995c
<pre>
Add about: scheme handler.
<a href="https://bugs.webkit.org/show_bug.cgi?id=288910">https://bugs.webkit.org/show_bug.cgi?id=288910</a>
<a href="https://rdar.apple.com/145910285">rdar://145910285</a>

Reviewed by Chris Dumez.

Currently about: scheme is handled ad-hoc independently. Add AboutSchemeHandler to support generic way to handle
about: scheme for each platform way.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::maybeLoadEmpty):
(WebCore::DocumentLoader::shouldCancelLoadingAboutURL const):
(WebCore::shouldCancelLoadingAboutURL): Deleted.
* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::isHandledByAboutSchemeHandler const):
(WebCore::DocumentLoader::setIsHandledByAboutSchemeHandler):
* Source/WebCore/loader/FrameLoadRequest.h:
(WebCore::FrameLoadRequest::isHandledByAboutSchemeHandler const):
(WebCore::FrameLoadRequest::setIsHandledByAboutSchemeHandler):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::load):
* Source/WebCore/platform/LegacySchemeRegistry.cpp:
(WebCore::LegacySchemeRegistry::registerURLSchemeAsHandledBySchemeHandler):
(WebCore::LegacySchemeRegistry::registerURLSchemeAsCORSEnabled):
* Source/WebKit/Shared/LoadParameters.h:
* Source/WebKit/Shared/LoadParameters.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/AboutSchemeHandler.cpp: Added.
(WebKit::AboutSchemeHandler::singleton):
(WebKit::AboutSchemeHandler::AboutSchemeHandler):
(WebKit::AboutSchemeHandler::handlerForURL const):
(WebKit::AboutSchemeHandler::platformStartTask):
(WebKit::AboutSchemeHandler::canHandleURL const):
(WebKit::AboutSchemeHandler::platformInitialize):
(WebKit::AboutSchemeHandler::registerHandler):
* Source/WebKit/UIProcess/AboutSchemeHandler.h: Added.
* Source/WebKit/UIProcess/Cocoa/AboutSchemeHandlerCocoa.mm: Added.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::m_pageForTesting):
(WebKit::WebPageProxy::loadRequestWithNavigationShared):
(WebKit::WebPageProxy::loadSimulatedRequest):
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::WebLoaderStrategy::tryLoadingUsingURLSchemeHandler):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::loadRequest):

Canonical link: <a href="https://commits.webkit.org/291957@main">https://commits.webkit.org/291957@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee7dd479f66eebddb066600499bd11c5683b6197

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94156 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13743 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3493 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99167 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44683 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96206 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14043 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22173 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71821 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29161 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97158 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10408 "Found 1 new test failure: fast/forms/ios/focus-input-via-button-no-scaling.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85001 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52162 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10099 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2690 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44001 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80332 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2780 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101209 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21208 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15426 "Found 2 new test failures: imported/w3c/IndexedDB-private-browsing/idbcursor_advance_index8.html imported/w3c/IndexedDB-private-browsing/idbcursor_continue_index8.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80825 "Found 29 new test failures: imported/w3c/web-platform-tests/css/css-masking/clip-path/clip-path-descendant-text-mutated-001.html imported/w3c/web-platform-tests/css/css-text/hyphens/hyphens-vs-float-clearance-002.html imported/w3c/web-platform-tests/css/css-transforms/animation/transform-non-invertible-discrete-interpolation.html imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-on-non-root.html imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html imported/w3c/web-platform-tests/css/css-view-transitions/massive-element-left-of-viewport-offscreen-new.html imported/w3c/web-platform-tests/css/css-view-transitions/massive-element-on-top-of-viewport-offscreen-new.html imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-with-classes-exit.html imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-with-classes-match-ident.html imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-with-classes-match-multiple-wildcard.html ... (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21460 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81018 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80207 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20068 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24752 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2108 "Found 2 new test failures: svg/W3C-I18N/tspan-dirRTL-ubEmbed-in-default-context.svg svg/W3C-I18N/tspan-dirRTL-ubEmbed-in-ltr-context.svg (failure)") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14370 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21192 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26371 "Found 1 new failure in UIProcess/RemotePageProxy.cpp") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20879 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24339 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22620 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/AboutSchemeHandler.cpp:
```diff
@@ -0,0 +1,83 @@
+    if (requestURL.hasOpaquePath()) {
+        auto opaquePath = requestURL.path();
+
+        if (auto it = m_handlers.find(opaquePath.toStringWithoutCopying()); it != m_handlers.end()) {
```
Can use `find<StringViewHashTranslator>(opaquePath)`

cdumez replied:
Also, why don't we use `get()` and null check the result?

basuke replied:
Because I didn't know that! Thanks!

cdumez commented on Source/WebKit/UIProcess/AboutSchemeHandler.cpp:
```diff
@@ -0,0 +1,83 @@
+bool AboutSchemeHandler::canHandleURL(const URL& url) const
+{
+    auto opaquePath = url.path();
+    return url.hasOpaquePath() && m_handlers.contains(opaquePath.toStringWithoutCopying());
```
Can use StringViewHashTranslator

cdumez commented on Source/WebCore/loader/FrameLoader.h:
```diff
@@ -358,6 +358,10 @@ class FrameLoader final : public CanMakeWeakPtr<FrameLoader> {
 
     void updateURLAndHistory(const URL&, RefPtr<SerializedScriptValue>&& stateObject, NavigationHistoryBehavior = NavigationHistoryBehavior::Replace);
 
+    bool canHandleAboutURL() const { return m_canHandleAboutURL; }
```
Maybe we can this `isHandledByAboutSchemeHandler()`?

cdumez commented on Source/WebKit/UIProcess/AboutSchemeHandler.h:
```diff
@@ -0,0 +1,60 @@
+class AboutSchemeHandler final : public WebURLSchemeHandler {
+    friend MainThreadNeverDestroyed<AboutSchemeHandler>;
+public:
+    class OpaquePathHandler {
```
Seems this is missing `WTF_MAKE_FAST_ALLOCATED`

cdumez replied:
It looks like this wasn't addressed?

cdumez commented on Source/WebKit/UIProcess/AboutSchemeHandler.h:
```diff
@@ -0,0 +1,60 @@
+
+    static AboutSchemeHandler& singleton();
+
+    void registerHandler(const String& opaquePath, std::unique_ptr<OpaquePathHandler>&&);
```
Does this need to be public?

basuke replied:
Subclass of OpaquePathHandler need to call this.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -913,6 +914,9 @@ WebPageProxy::WebPageProxy(PageClient& pageClient, WebProcessProxy& process, Ref
         gpuProcess->setPresentingApplicationAuditToken(process.coreProcessIdentifier(), m_webPageID, m_presentingApplicationAuditToken);
 #endif
+
+    Ref aboutSchemeHandler = AboutSchemeHandler::singleton();
```
We shouldn't need `Ref` here given that this is a singleton. The static analyzer should be able to recognize this. I don't even think we need a local variable here.

cdumez commented on Source/WebCore/loader/FrameLoader.h:
```diff
@@ -543,6 +547,7 @@ class FrameLoader final : public CanMakeWeakPtr<FrameLoader> {
 
     bool m_errorOccurredInLoading { false };
     bool m_doNotAbortNavigationAPI { false };
+    bool m_isHandledByAboutSchemeHandler { false };
```
I doubt this is the right location to be storing this flag. Normally, flags that have to do with one particular navigation are stored on the DocumentLoader, not the FrameLoader. The reason is that the FrameLoader gets re-used for every navigation, while the DocumentLoader is per navigation.

cdumez replied:
You may want to look at the `DocumentLoader::m_isRequestFromClientOrUserInput` flag for example, I think it can also come from the UIProcess like your new flag.

cdumez commented on Source/WebKit/UIProcess/AboutSchemeHandler.cpp:
```diff
@@ -0,0 +1,82 @@
+    if (requestURL.hasOpaquePath()) {
+        auto opaquePath = requestURL.path();
+
+        if (const auto& handler = m_handlers.get<StringViewHashTranslator>(opaquePath)) {
```
just `auto`, we don't normally use `const` for variables in WebKit and the reference here is not helpful.

cdumez commented on Source/WebKit/UIProcess/AboutSchemeHandler.cpp:
```diff
@@ -0,0 +1,82 @@
+void AboutSchemeHandler::platformStartTask(WebPageProxy&, WebURLSchemeTask& task)
+{
+    auto requestURL = task.request().url();
+    if (requestURL.hasOpaquePath()) {
```
Can we do an early return instead? We prefer early returns in general in WebKit to avoid nesting.

cdumez commented on Source/WebKit/UIProcess/AboutSchemeHandler.cpp:
```diff
@@ -0,0 +1,82 @@
+        auto opaquePath = requestURL.path();
+
+        if (const auto& handler = m_handlers.get<StringViewHashTranslator>(opaquePath)) {
+            handler->loadContent(requestURL, [protectedTask = Ref { task }](auto&& response, auto&& buffer) mutable {
```
Just `task = Ref { Task }`, no need for the `protected` prefix which we try to avoid in new code.

cdumez commented on Source/WebKit/UIProcess/AboutSchemeHandler.cpp:
```diff
@@ -0,0 +1,82 @@
+
+bool AboutSchemeHandler::canHandleURL(const URL& url) const
+{
+    return url.hasOpaquePath() && m_handlers.contains<StringViewHashTranslator>(url.path());
```
Shouldn't this check if the URL has the `about` protocol? Seems like the most basic requirement for this class.

cdumez commented on Source/WebKit/UIProcess/AboutSchemeHandler.cpp:
```diff
@@ -0,0 +1,82 @@
+void AboutSchemeHandler::platformInitialize()
+{
+#if PLATFORM(COCOA) && HAVE(CUSTOM_ABOUT_SCHEME_HANDLER)
+    registerCocoaAboutHandlers(this);
```
Can `registerCocoaAboutHandlers()` ever be called with nullptr? If not, we should pass `*this`

cdumez commented on Source/WebKit/UIProcess/AboutSchemeHandler.h:
```diff
@@ -0,0 +1,61 @@
+    void registerHandler(const String& opaquePath, std::unique_ptr<OpaquePathHandler>&&);
+    bool canHandleURL(const URL&) const;
+
+    constexpr static auto scheme = "about"_s;
```
We usually use `static constexpr`, not the other way around.

cdumez commented on Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
```diff
@@ -292,6 +292,10 @@ bool WebLoaderStrategy::tryLoadingUsingURLSchemeHandler(ResourceLoader& resource
         return false;
 
+    auto url = resourceLoader.request().url();
+    if (url.isAboutBlank() || url.isAboutSrcDoc())
```
We shouldn't need this code here. If this is ever possible for this condition to be true, then it means you are missing some checks in WebCore/

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -2186,6 +2190,7 @@ RefPtr<API::Navigation> WebPageProxy::loadFile(const String& fileURLString, cons
     loadParameters.userData = UserData(legacyMainFrameProcess().transformObjectsToHandles(userData).get());
     loadParameters.publicSuffix = WebCore::PublicSuffixStore::singleton().publicSuffix(loadParameters.request.url());
     loadParameters.isRequestFromClientOrUserInput = isAppInitiated;
+    loadParameters.isHandledByAboutSchemeHandler = AboutSchemeHandler::singleton().canHandleURL(loadParameters.request.url());
```
We're in `loadFile()`, why is this needed? Surely a file URL cannot be handled by the about scheme handler?

cdumez commented on Source/WebCore/loader/DocumentLoader.h:
```diff
@@ -525,6 +525,10 @@ class DocumentLoader
 
+    bool isHandledByAboutSchemeHandler() const { return m_isHandledByAboutSchemeHandler; }
+    void setIsHandledByAboutSchemeHandler(bool isHandledByAboutSchemeHandler) { m_isHandledByAboutSchemeHandler = isHandledByAboutSchemeHandler; }
+    bool shouldCancelLoadingAboutURL(const URL&) const;
```
Why is this public?

cdumez commented on Source/WebKit/UIProcess/AboutSchemeHandler.h:
```diff
@@ -0,0 +1,62 @@
+
+class AboutSchemeHandler final : public WebURLSchemeHandler {
+    WTF_MAKE_FAST_ALLOCATED;
+    friend MainThreadNeverDestroyed<AboutSchemeHandler>;
```
We usually put those after the `private:`

cdumez replied:
This comment was marked as resolved but not actually addressed.

cdumez commented on Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
```diff
@@ -292,6 +292,12 @@ bool WebLoaderStrategy::tryLoadingUsingURLSchemeHandler(ResourceLoader& resource
 
+    if (RefPtr documentLoader = resourceLoader.documentLoader()) {
+        auto url = resourceLoader.request().url();
+        if (url.protocolIsAbout() && !documentLoader->isHandledByAboutSchemeHandler())
```
Why do we need this?

basuke replied:
This test `TestWebKitAPI.WebKit.AttrStyle` fails without this check.

cdumez replied:
But the question is why is it needed? Why isn't the check in WebCore sufficient?

cdumez replied:
Oh, I think I get it, the check in WebCore only covers main resource loads. The issue is that `TestWebKitAPI.WebKit.AttrStyle` is trying to do a subresource load to an `about:` URL..

cdumez replied:
This makes me think that there is a bug with the about scheme handler. It should be harmless to go to the about scheme handler, it should respond with an empty response and it should not have a different behavior than if we had no about scheme handler.

Can we investigate why the behavior is different? Maybe calling `didComplete` without calling `didReceiveResponse` is wrong?

basuke replied:
Got it. I'll dig into this.

basuke replied:
We need to pass ResourceError do didComplete If we didn't send any response.

cdumez commented on Source/WebKit/UIProcess/AboutSchemeHandler.h:
```diff
@@ -0,0 +1,64 @@
+    static constexpr auto scheme = "about"_s;
+
+private:
+    WTF_MAKE_FAST_ALLOCATED;
```
This is in the wrong class, it should come right after:
```
class AboutSchemeHandler final : public WebURLSchemeHandler {
```

You were supposed to move the `friend` declaration, not this.

cdumez reviewed (approved):
LGTM with nits.

cdumez commented on Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
```diff
@@ -292,6 +292,9 @@ bool WebLoaderStrategy::tryLoadingUsingURLSchemeHandler(ResourceLoader& resource
     if (!webPage || !webFrame)
         return false;
 
+    if (resourceLoader.request().url().protocolIsAbout() && is<SubresourceLoader>(resourceLoader))
```
Is this still needed?

basuke replied:
It seems we should keep this as is. I tried several solution that handling about:blank, returning empty 200 response to all request or returning error, some tests are still failing.

basuke committed:
Add about: scheme handler.
https://bugs.webkit.org/show_bug.cgi?id=288910
rdar://145910285

Reviewed by Chris Dumez.

Currently about: scheme is handled ad-hoc independently. Add AboutSchemeHandler to support generic way to handle
about: scheme for each platform way.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::maybeLoadEmpty):
(WebCore::DocumentLoader::shouldCancelLoadingAboutURL const):
(WebCore::shouldCancelLoadingAboutURL): Deleted.
* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::isHandledByAboutSchemeHandler const):
(WebCore::DocumentLoader::setIsHandledByAboutSchemeHandler):
* Source/WebCore/loader/FrameLoadRequest.h:
(WebCore::FrameLoadRequest::isHandledByAboutSchemeHandler const):
(WebCore::FrameLoadRequest::setIsHandledByAboutSchemeHandler):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::load):
* Source/WebCore/platform/LegacySchemeRegistry.cpp:
(WebCore::LegacySchemeRegistry::registerURLSchemeAsHandledBySchemeHandler):
(WebCore::LegacySchemeRegistry::registerURLSchemeAsCORSEnabled):
* Source/WebKit/Shared/LoadParameters.h:
* Source/WebKit/Shared/LoadParameters.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/AboutSchemeHandler.cpp: Added.
(WebKit::AboutSchemeHandler::singleton):
(WebKit::AboutSchemeHandler::AboutSchemeHandler):
(WebKit::AboutSchemeHandler::handlerForURL const):
(WebKit::AboutSchemeHandler::platformStartTask):
(WebKit::AboutSchemeHandler::canHandleURL const):
(WebKit::AboutSchemeHandler::platformInitialize):
(WebKit::AboutSchemeHandler::registerHandler):
* Source/WebKit/UIProcess/AboutSchemeHandler.h: Added.
* Source/WebKit/UIProcess/Cocoa/AboutSchemeHandlerCocoa.mm: Added.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::m_pageForTesting):
(WebKit::WebPageProxy::loadRequestWithNavigationShared):
(WebKit::WebPageProxy::loadSimulatedRequest):
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::WebLoaderStrategy::tryLoadingUsingURLSchemeHandler):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::loadRequest):

Canonical link: https://commits.webkit.org/291957@main
